cranelift/interpreter/src/step.rs

```diff
@@ -0,0 +1,717 @@
+                IntCC::UnsignedGreaterThanOrEqual => Value::ge(
+                    &$a.clone().convert(ValueConversionKind::ToUnsigned)?,
+                    &$b.clone().convert(ValueConversionKind::ToUnsigned)?,
+                )?,
+                IntCC::UnsignedLessThan => Value::lt(
+                    &$a.clone().convert(ValueConversionKind::ToUnsigned)?,
+                    &$b.clone().convert(ValueConversionKind::ToUnsigned)?,
+                )?,
+                IntCC::UnsignedLessThanOrEqual => Value::le(
+                    &$a.clone().convert(ValueConversionKind::ToUnsigned)?,
+                    &$b.clone().convert(ValueConversionKind::ToUnsigned)?,
+                )?,
+                IntCC::Overflow => unimplemented!("IntCC::Overflow"),
+                IntCC::NotOverflow => unimplemented!("IntCC::NotOverflow"),
+            }
+        };
+    }
+    // TODO should not have to use signed less-than/greater-than
+    macro_rules! fcmp {
+        () => {
+            fcmp!(inst.fp_cond_code().unwrap(), &arg!(0), &arg!(1))
+        };
+        ($f: expr, $a: expr, $b: expr) => {
+            match $f {
+                FloatCC::Ordered => Value::eq($a, $b)? || Value::lt($a, $b)? || Value::gt($a, $b)?,
+                FloatCC::Unordered => Value::uno($a, $b)?,
+                FloatCC::Equal => Value::eq($a, $b)?,
+                FloatCC::NotEqual => {
+                    Value::lt($a, $b)? || Value::gt($a, $b)? || Value::uno($a, $b)?
+                }
+                FloatCC::OrderedNotEqual => Value::lt($a, $b)? || Value::gt($a, $b)?,
+                FloatCC::UnorderedOrEqual => Value::eq($a, $b)? || Value::uno($a, $b)?,
+                FloatCC::LessThan => Value::lt($a, $b)?,
+                FloatCC::LessThanOrEqual => Value::lt($a, $b)? || Value::eq($a, $b)?,
+                FloatCC::GreaterThan => Value::gt($a, $b)?,
+                FloatCC::GreaterThanOrEqual => Value::gt($a, $b)? || Value::eq($a, $b)?,
+                FloatCC::UnorderedOrLessThan => Value::uno($a, $b)? || Value::lt($a, $b)?,
+                FloatCC::UnorderedOrLessThanOrEqual => {
+                    Value::uno($a, $b)? || Value::lt($a, $b)? || Value::eq($a, $b)?
+                }
+                FloatCC::UnorderedOrGreaterThan => Value::uno($a, $b)? || Value::gt($a, $b)?,
+                FloatCC::UnorderedOrGreaterThanOrEqual => {
+                    Value::uno($a, $b)? || Value::gt($a, $b)? || Value::eq($a, $b)?
+                }
+            }
+        };
+    }
+    fn sum<V: Value>(head: V, tail: SmallVec<[V; 1]>) -> Result<i64, ValueError> {
+        let mut acc = head;
+        for t in tail {
+            acc = Value::add(acc, t)?;
+        }
+        acc.into_int()
+    }
+
+    Ok(match inst.opcode() {
+        Opcode::Jump | Opcode::Fallthrough => ControlFlow::ContinueAt(branch!(), args!()),
+        Opcode::Brz => branch_when!(!arg!(0).into_bool()?),
+        Opcode::Brnz => branch_when!(arg!(0).into_bool()?),
```

Comment:
Ok, I replaced the macros with closures in 5a121dd. I like the idea of closures better than macros but it made a few things a bit harder to read so I left it as a separate commit in case we want to revert.